### PR TITLE
Active forces with manifold constraints

### LINKED ISSUE #869: Remove oneD, sphere, and ellipsoid constraints
## Description

Remove constrain.oneD, constrain.sphere, and the ellipsoid active matter constraint.

## Motivation and context

These are all being replaced with integrators that support constraints on generic manifolds.

### PR DESCRIPTION
**Pull request by SchoeniPhlippsn**
## Description
Overhaul of active forces. I integrated manifolds to allow for active particles on curved surfaces and deleted the original ellipsoidal technique. Therefore, I divided the active force class into two separate classes ( constraint: `ActiveForceConstraintCompute` and unconstrained: `ActiveForceCompute`). All the manifold constraints can be used. 

## Motivation and context

The active force class contained old, unused code from the outdated ellipsodal constraints. By switching to manifold constraint, users can restrict active motion to more manifolds and apply active forces in combination with RATTLE integrators. 

Resolves  #869 and #936

## How has this been tested?

Pytests  have been updated and extended. I also ckecked the documentation.

## Change log

```
Adding manifold constraints to active forces.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/ActiveForceConstraintCompute.h
Comment (j-proc) [inline hoomd/md/ActiveForceConstraintCompute.h:123]: I'm not sue if it's necessary to check this here as I believe it will work out to once per timestep. This might be a very cheap operation, but assuming it is not, you could perhaps check it once when the manifold is set and then again each time the box is changed, assuming the manifold does not change with it. You can do this by subscribing the function to the box change signal.

File: hoomd/md/ActiveForceConstraintCompute.h
Comment (j-proc) [inline hoomd/md/ActiveForceConstraintCompute.h:125]: Most errors in HOOMD are done with the m_exec_conf objects. I believe this is to properly handle a variety of run environments that the std library may not.

File: hoomd/md/force.py
Comment (j-proc) [inline hoomd/md/force.py:378]: I wonder if it would make sense/be plausible to remove and recreate the associated objects post construction?

File: hoomd/md/CMakeLists.txt
Comment (j-proc) [inline hoomd/md/CMakeLists.txt:54]: I think you may want to add a few more of your files to the various sources.

**Review by j-proc — COMMENTED** [review]
In general this seems like a major cleanup of active forces while simultaneously generalizing the constraints. I had some minor comments but I think with your feedback on those and the style corrections of pre-commit this looks ready to go.

File: hoomd/md/ActiveForceConstraintCompute.h
Comment (joaander) [reply hoomd/md/ActiveForceConstraintCompute.h:125]: The use of `std::runtime_error` is correct here.

Exceptions are handled properly even in MPI and provide a cleaner message to the user. There is another big cleanup project in the future to convert `error << msg; throw runtime_error` to `throw runtime_error(msg)` in all of hoomd.

**Review by joaander — CHANGES_REQUESTED** [review]
In addition to some specific change requests, I wonder how these changes impact #898. We do need to switch to using an Updater to schedule the rotational diffusion updates, but with this PR those updates can now depend on the Manifold.

@b-butler what do you think? Should we move the rotational diffusion out into a completely separate class which would need to duplicate some of the manifold setup, or should we make an Updater that links to an ActiveForceCompute and leverages the existing code in these classes to implement the diffusion?

We don't have to implement #898 in this PR - but let's have a plan before we merge this one. We do need to eventually fix the issue because there are a number of issues with calling the rotational diffusion during the force computation.

File: hoomd/md/force.py
Comment (SchoeniPhlippsn) [reply hoomd/md/force.py:378]: Usually it is not a good idea to switch constraints after initialisation as the forces become discontinuous. The only thing I can imagine, users might be interested in, is starting with a constraint and lifting it later. But this means detatching the restricted active force from the simulationa and attaching the unrestricted one afterwards anyway.

**Comment by SchoeniPhlippsn** [comment]
Sorry that it took some time. However, this is open for review again.

File: hoomd/md/CMakeLists.txt
Comment (SchoeniPhlippsn) [reply hoomd/md/CMakeLists.txt:51]: Which ones did I miss?

**Review by j-proc — APPROVED** [review]
This looks good to go to me. The minor possible changes can always be made enhancement issues but no reason to hold this up anymore.

**Comment by j-proc** [comment]
Merge conflicts actually look a bit worse than I originally anticipated, I had thought I might help out. A rebase might be in order to clean it up, but outside the simple parts where both changes need to be removed, it seems like things @SchoeniPhlippsn should probably handle to make sure it stays correct.

File: hoomd/md/ActiveForceConstraintCompute.h
Comment (joaander) [inline hoomd/md/ActiveForceConstraintCompute.h:172]:
```suggestion
            Scalar phi = slow::atan(dot_prod * dot_perp_prod);
```
See below.

File: hoomd/md/ActiveForceConstraintCompute.h
Comment (joaander) [inline hoomd/md/ActiveForceConstraintCompute.h:127]:
```suggestion
        quat<Scalar> rot_quat = quat<Scalar>::fromAxisAngle(norm, theta);
```
Simplify.

File: hoomd/md/ActiveForceConstraintCompute.h
Comment (joaander) [inline hoomd/md/ActiveForceConstraintCompute.h:184]:
```suggestion
            quat<Scalar> rot_quat = quat<Scalar>::(rot_vec, phi);
```
Simplify.

File: hoomd/md/ActiveForceConstraintComputeGPU.cuh
Comment (joaander) [inline hoomd/md/ActiveForceConstraintComputeGPU.cuh:85]: Simplify this with the same changes as the CPU code.

File: hoomd/md/ActiveForceConstraintComputeGPU.cuh
Comment (joaander) [inline hoomd/md/ActiveForceConstraintComputeGPU.cuh:148]: Simplify this with the same changes as the CPU code.

File: hoomd/md/ActiveForceConstraintComputeGPU.h
Comment (joaander) [inline hoomd/md/ActiveForceConstraintComputeGPU.h:43]: Use an Autotuner in each of `ActiveForceComputeGPU` and `ActiveForceConstraintComputeGPU<Manifold>`. Different GPUs, system sizes, etc... lead to different optimal block sizes. `Autotuner` finds the fastest block size.

File: hoomd/md/CMakeLists.txt
Comment (joaander) [inline hoomd/md/CMakeLists.txt:49]: Add `ActiveForceConstraintComputeGPU.cuh` to this list.

File: hoomd/md/force.py
Comment (joaander) [inline hoomd/md/force.py:390]: Implement the manifold active force compute in a separate class to avoid issues like #998 for the integration methods.

File: hoomd/md/force.py
Comment (joaander) [inline hoomd/md/force.py:294]:
```suggestion
            orientation of a particle :math:`[\mathrm{force} \cdot \mathrm{length}]`. It is defined per particle type and stays
```
Define units of the `active_torque` parameter. The text in this section will need to be wrapped appropriately after making this change.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks for the updates. I have a few suggestions to simplify the vector math code and make it more readable by other developers, we need to use block size autotuners to improve performance, there is a missing header in the CMake script, and the active force manifold class needs to be separated from the non-manifold active force.

File: hoomd/md/force.py
Comment (joaander) [inline hoomd/md/force.py:377]: I'm going to rename this class name to `ActiveOnManifold`. `ActiveConstraint` suggests that it is the constraint itself that is somehow active.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**
